hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/tagdetails/AbstractTagToken.java

```diff
@@ -9,60 +9,42 @@
 package org.eclipse.hawkbit.ui.common.tagdetails;
 
 import java.io.Serializable;
-import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
 
 import org.eclipse.hawkbit.repository.model.BaseEntity;
+import org.eclipse.hawkbit.repository.model.NamedEntity;
 import org.eclipse.hawkbit.ui.SpPermissionChecker;
 import org.eclipse.hawkbit.ui.common.table.BaseEntityEventType;
 import org.eclipse.hawkbit.ui.common.table.BaseUIEntityEvent;
+import org.eclipse.hawkbit.ui.common.tagdetails.TagPanelLayout.TagAssignmentListener;
 import org.eclipse.hawkbit.ui.management.state.ManagementUIState;
-import org.eclipse.hawkbit.ui.utils.SPUIDefinitions;
-import org.eclipse.hawkbit.ui.utils.SPUIStyleDefinitions;
 import org.eclipse.hawkbit.ui.utils.UINotification;
 import org.eclipse.hawkbit.ui.utils.VaadinMessageSource;
 import org.vaadin.spring.events.EventBus;
 import org.vaadin.spring.events.EventBus.UIEventBus;
-import org.vaadin.tokenfield.TokenField;
-import org.vaadin.tokenfield.TokenField.InsertPosition;
-
-import com.vaadin.data.Container;
-import com.vaadin.data.Item;
-import com.vaadin.data.Property;
-import com.vaadin.data.util.IndexedContainer;
-import com.vaadin.server.FontAwesome;
-import com.vaadin.shared.ui.combobox.FilteringMode;
-import com.vaadin.ui.Button;
-import com.vaadin.ui.CssLayout;
+
 import com.vaadin.ui.UI;
-import com.vaadin.ui.themes.ValoTheme;
 
 /**
  * Abstract class for target/ds tag token layout.
  *
  * @param <T>
  *            the special entity
  */
-public abstract class AbstractTagToken<T extends BaseEntity> implements Serializable {
-
-    private static final String ID_PROPERTY = "id";
-    private static final String NAME_PROPERTY = "name";
-    private static final String COLOR_PROPERTY = "color";
+public abstract class AbstractTagToken<T extends BaseEntity> implements Serializable, TagAssignmentListener {
 
     protected static final int MAX_TAG_QUERY = 500;
 
     private static final long serialVersionUID = 6599386705285184783L;
 
-    protected TokenField tokenField;
-
-    protected IndexedContainer container;
-
-    protected final transient Map<Long, TagData> tagDetails = new ConcurrentHashMap<>();
-
-    protected final transient Map<Long, TagData> tokensAdded = new HashMap<>();
+    protected TagPanelLayout tagPanelLayout;
 
-    protected CssLayout tokenLayout = new CssLayout();
+    protected final transient Map<Long, TagData> tagDetailsById = new ConcurrentHashMap<>();
```

Comment:
Do you really need both tagDetailsById and tagDetailsByName? Maybe it would be better to use Map.Entry<Long, String> as a key...